proposals/js-string-builtins/Overview.md

```diff
@@ -28,221 +28,277 @@ The first problem is that existing API’s require a calling convention conversi
+When the module is instantiated, a unique instantiation of the builtins are created. This means that re-exports of builtin functions will have different identities if they come from different instances. This is a useful property for future extensions to bind memory to builtins or evolve the types as things like type-imports or a core stringref type are added (see below).
 
 ## Feature detection
 
-Users may wish to detect if a specific builtin module is available in their system.
+Users may wish to detect if a specific builtin is available in their system.
 
-A simple option is to add `WebAssembly.hasBuiltinModule(name)` method. This is likely too coarse grained though, users may wish to know if a specific function in a builtin module is available, as new ones may be added over time.
-
-A more general option would then be to extend `WebAssembly.validate` to also take a list of builtin modules to enable, like compile does. After validating the module, the eager link checking that compile does is also performed. This would allow checking for the presence of individual parts of a builtin module.
+For this purpose, `WebAssembly.validate` is extended to take a list of builtins to enable, like compile does. After validating the module, the eager link checking that compile does is also performed. Users can inspect the result of validate on modules importing builtins to see if they are supported.
 
 ## Polyfilling
 
-If a user wishes to polyfill these imports for some reason, or is running on a system without a builtin module, these imports may be provided as normal through instantiation.
+If a user wishes to polyfill these imports for some reason, or is running on a system without a builtin, these imports may be provided as normal through instantiation.
 
-## JS String Builtin API
+## UTF8/WTF8 support
+
+As stated above in 'goals for builtins', builtins are intended to just wrap existing primitives and not invent new functionality.
 
-The following is an initial set of function builtins for JavaScript String. The builtin module name is `js-string`. Each example includes pseudo-code illustrating their operation and some descriptive text.
+JS Strings are semantically a sequence of 16-bit code units (referred to as char codes in method naming), and there are no builtin operations on them to acquire a UTF-8 or WTF-8 view. This makes it difficult to write wasm builtins for these encodings without introducing significant new logic to them.
 
-TODO: formalize these better.
+There is however, the Encoding API for `TextEncoder`/`TextDecoder` which can be used for UTF-8 support. However, this is technically a separate spec from JS and may not be available on all JS engines (in practice it's available widely). This proposal exposes UTF-8 data conversions using this API under separate `wasm:text-encoder` `wasm:text-decoder` interfaces which are available when the host implements these interfaces.
```

Comment:
Hmm, good point. The tricky part about that use-case for WTF-8 is:
  1. There's no prior API's that operate on WTF-8 in the Web (AFAIK), it's a new encoding that would need to be adopted. It's also not really a web standard yet either, it's just a repo on github (http://simonsapin.github.io/wtf-8/). Adding it to this proposal would be a major new addition to the web platform, and would need to get consensus from other stakeholders that it's okay. If we got that, I'd feel better about it.
  2. The WTF-8 spec says '[WTF-8](http://simonsapin.github.io/wtf-8/#wtf-8) must not be used to represent text in a file format or for transmission over the Internet' and so supporting it for embedding string constants in a wasm binary seems counter to the intent of the proposal.

Is there some other way we could support this use-case? The TextDecoder interface does support ASCII/Latin-1 and so we could add support for that encoding on JS string creation path.

Reply:
Re 1, I'm not so worried about that part. WTF-8 is a small and straightforward relaxation of UTF-8. It's just a way to represent the JS strings that are already prevalent on the web in a 1-byte-codeunit format. Also, it's not like we're adding WTF-8 strings as a general-purpose primitive, nor are we encouraging anyone to use it as a data interchange format. I see this closer to e.g. `TextDecoder` supporting a (huge!) bunch of legacy encodings that aren't otherwise used on the web (note that we didn't have a string-to-wtf8 function, only string-from-wtf8).

Re 2, yes, that's an unresolved wrinkle that I'm also unhappy about. Wasm satisfies the "used internally in ~self-contained~ systems with components that need to support potentially ill-formed UTF-16 for legacy reasons" part, it just unfortunately also has a "transmission over the Internet" step between the module producer and the browser executing the module.

All that said, it's probably acceptable, at least for a first version, to have only UTF-8 and WTF-16 (aka "charCode") support. Since the relevant use case is string constants, toolchains can check the string for UTF-8 validity, and use the UTF-8 encoding most of the time, and fall back to WTF-16 in the rare cases where UTF-8 is too strict. We can reopen the WTF-8 discussion if and when evidence arises that the UTF-8+WTF-16 combo is insufficient.

I think having a way to create strings from ASCII data would be nice in general (for performance: one use case where it came up is number-to-string conversion). I'd like to see it added right away, but I admittedly don't have a strong reason to argue that it is necessary for an MVP.

Reply:
> 
> I think having a way to create strings from ASCII data would be nice in general (for performance: one use case where it came up is number-to-string conversion). I'd like to see it added right away, but I admittedly don't have a strong reason to argue that it is necessary for an MVP.

It's probably worth adding, I won't do it in this PR but in a future one.